pkg/innerring/processors/container/process_container.go

```
@@ -252,6 +254,137 @@ func (cp *Processor) approveDeleteContainer(e containerEvent.RemoveContainerRequ
 	}
 }
 
+func (cp *Processor) processSetAttributeRequest(req containerEvent.SetAttributeRequest) {
+	if !cp.alphabetState.IsAlphabet() {
+		cp.log.Info("non alphabet mode, ignore attribute setting")
+		return
+	}
+
+	id, err := cid.DecodeBytes(req.ID)
+	if err != nil {
+		cp.log.Error("attribute setting check failed",
+			zap.Error(fmt.Errorf("invalid container ID: %w", err)))
+		return
+	}
+	if id.IsZero() {
+		cp.log.Error("attribute setting check failed",
+			zap.Error(cid.ErrZero))
+		return
+	}
+
+	err = cp.checkSetAttributeRequest(req, id)
+	if err != nil {
+		cp.log.Error("attribute setting check failed",
+			zap.Stringer("container", id), zap.Error(err))
+		return
+	}
+
+	cp.approveSetAttributeRequest(req, id)
+}
+
+func (cp *Processor) checkSetAttributeRequest(req containerEvent.SetAttributeRequest, id cid.ID) error {
+	cnr, err := cp.cnrClient.Get(req.ID)
+	if err != nil {
+		return fmt.Errorf("get container by ID: %w", err)
+	}
+
+	signedData := client.GetSignedSetContainerAttributeParameters(client.SetContainerAttributeParameters{
+		ID:         id,
+		Attribute:  req.Attribute,
+		Value:      req.Value,
+		ValidUntil: time.Unix(req.ValidUntil, 0),
```

Check `ValidUntil` here as well, fail fast.